programs/local/LocalServer.cpp

```diff
@@ -191,335 +180,460 @@ static void attachSystemTables(ContextPtr context)
+        "<yandex>"
+        "    <profiles>"
+        "        <default></default>"
+        "    </profiles>"
+        "    <users>"
+        "        <default>"
+        "            <password></password>"
+        "            <networks>"
+        "                <ip>::/0</ip>"
+        "            </networks>"
+        "            <profile>default</profile>"
+        "            <quota>default</quota>"
+        "        </default>"
+        "    </users>"
+        "    <quotas>"
+        "        <default></default>"
+        "    </quotas>"
+        "</yandex>";
+
+    ConfigurationPtr users_config;
+
+    if (config().has("users_config") || config().has("config-file") || fs::exists("config.xml"))
     {
-        cleanup();
+        const auto users_config_path = config().getString("users_config", config().getString("config-file", "config.xml"));
+        ConfigProcessor config_processor(users_config_path);
+        const auto loaded_config = config_processor.loadConfig();
+        config_processor.savePreprocessedConfig(loaded_config, config().getString("path", DBMS_DEFAULT_PATH));
+        users_config = loaded_config.configuration;
     }
-    catch (...)
+    else
     {
-        tryLogCurrentException(__PRETTY_FUNCTION__);
+        users_config = getConfigurationFromXMLString(minimal_default_user_xml);
     }
 
-    std::cerr << getCurrentExceptionMessage(config().hasOption("stacktrace")) << '\n';
-
-    /// If exception code isn't zero, we should return non-zero return code anyway.
-    return e.code() ? e.code() : -1;
+    if (users_config)
+        global_context->setUsersConfig(users_config);
+    else
+        throw Exception("Can't load config for users", ErrorCodes::CANNOT_LOAD_CONFIG);
 }
 
 
-std::string LocalServer::getInitialCreateTableQuery()
+bool LocalServer::processMultiQuery(const String & all_queries_text)
 {
-    if (!config().has("table-structure"))
-        return {};
+    auto process_single_query = [&](const String & query_to_execute, const String &, ASTPtr)
+    {
+        try
+        {
+            processSingleQueryImpl(query_to_execute, query_to_execute, nullptr, echo_queries, false);
+        }
+        catch (...)
+        {
+            local_server_exception = std::make_unique<Exception>(getCurrentExceptionMessage(false), getCurrentExceptionCode());
+            have_error = true;
+        }
+    };
 
-    auto table_name = backQuoteIfNeed(config().getString("table-name", "table"));
-    auto table_structure = config().getString("table-structure");
-    auto data_format = backQuoteIfNeed(config().getString("table-data-format", "TSV"));
-    String table_file;
-    if (!config().has("table-file") || config().getString("table-file") == "-") /// Use Unix tools stdin naming convention
-        table_file = "stdin";
-    else /// Use regular file
-        table_file = quoteString(config().getString("table-file"));
+    return processMultiQueryImpl(all_queries_text, process_single_query);
```

Comment:
processMultiQuery -> processMultiQueryImpl -> process_single_query -> processSingleQueryImpl, right? A virtual method calls a base method parameterized by a lambda callback, that in turn calls another base method. I'll try to think of some simpler control flow.

Reply:
> processMultiQuery -> processMultiQueryImpl -> process_single_query -> processSingleQueryImpl, right?

Yep.